Refactor payment methods components

Aljullu opened the pull request:
This PR includes several refactors I did as part of fixing #3166 (I will open follow-up PRs with actual fixes tomorrow). I wanted to extract refactors to a separate PR so they are easier to review and don't get blocked by other changes.

Basically, the rationale of the changes in this PR is:
* Make components smaller so they are easier to work with.
* Remove some `useRef()` usages that didn't seem to be necessary anymore.
* Simplify some of the logic along the way.

### How to test the changes in this Pull Request:

This PR doesn't include any behavior change, so everything should keep working as before.
1. Go to the Checkout block and interact with the payment method tabs. Verify you can toggle them and tab contents always appear.
2. Finish a purchase with one of the tabs and verify the order was done with the correct payment method.

It would be good to test that there are no performance regressions with `trunk`:
* With the devtools Network tab, verify there are no more calls to Stripe API than there were in `trunk`.
* Feel free to add some `console.log()` in some components to verify there are not excessive re-renders compared to `trunk`.

nerrad commented on assets/js/base/components/payment-methods/new-payment-method-options.js:
```diff
@@ -0,0 +1,82 @@
+import Tabs from '../tabs';
+import NewPaymentMethodTab from './new-payment-method-tab';
+
+const NewPaymentMethodOptions = () => {
```
Please add some jsdocs here.

nerrad commented on assets/js/base/components/payment-methods/new-payment-method-tab.js:
```diff
@@ -0,0 +1,45 @@
+ */
+import PaymentMethodErrorBoundary from './payment-method-error-boundary';
+
+const NewPaymentMethodTab = ( { children, allowsSaving } ) => {
```
Nice splitting this out into its own component.

Please add some jsdocs here.

nerrad reviewed (approved):
Really great work here @Aljullu 👏🏻 - this cleans up the code nicely and reduces non-needed refs. I have a couple comments about jsdocs.

Also why the `new-` and `New` prefix for the new components? I'd be fine with just dropping it?

I tested this fairly thoroughly including testing Apple Pay on my [test site](https://ephemeral-nerrad-20201002.atomicsites.blog) with this branch.

I also tested for possible regressions with bugs I've encountered before but it's all looking good.

Pre-approving, however I would like to see my comments addressed re dropping the `new` prefix and adding jsdocs.

Aljullu commented:
Thanks for the review @nerrad!

> Also why the `new-` and `New` prefix for the new components? I'd be fine with just dropping it?

The idea was to differentiate `New` from `Saved` payment methods. But dropping `new` makes sense since that's the default for all payment methods. Done in 600dd0cdae652036d5732c7164fef9ff50bfe592.

> jsdocs

Added in 54f4f919621bac2757efde36d9f8edc2b67e4989. I also added propTypes when necessary in 122083398463dce1443911a9185b26bc94ac0eea.
